python/src/iceberg/io/base.py

```diff
@@ -0,0 +1,139 @@
+
+    @abstractmethod
+    def __len__(self):
+        """Returns the total length of the file, in bytes"""
+
+    @property
+    def location(self):
+        """The fully-qualified location of the input file"""
+        return self._location
+
+    @property
+    @abstractmethod
+    def exists(self):
+        """Checks whether the file exists"""
+
+    @abstractmethod
+    def __enter__(self):
+        """Enter context for InputFile
+
+        This method should assign a seekable stream to `self.input_stream` and
+        return `self`. If the file does not exist, a FileNotFoundError should
+        be raised. To ensure that opening the file is logged, call
+        `super().__enter__()` at the beginning of the implementation of this
+        method."""
+
+        log.info(f"Opening file {self._location}")
+
+    @abstractmethod
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        """Exit context for InputFile
+
+        This method should perform any necessary teardown. To ensure that closing the
+        file is logged, call `super().__exit__()` at the beginning of the implementation
+        of this method."""
+
+        log.info(f"Closing file {self._location}")
+
+
+class OutputFile(ABC):
+    """A base class for OutputFile implementations"""
+
+    def __init__(self, location: str):
+        self._location = location
+
+    @abstractmethod
+    def __call__(self, file_obj, overwrite: bool = False, **kwargs):
+        """Set a file-like object, an overwrite flag, and any other kwargs before
+        opening a context manager. Exiting the context manager should remove these where
+        applicable."""
+        self._file_obj = file_obj
+        self._overwrite = overwrite
+
+    @abstractmethod
+    def __len__(self):
+        """Returns the total length of the file, in bytes"""
+
+    @property
+    def location(self):
+        """The fully-qualified location of the output file"""
+        return self._location
+
+    @property
+    @abstractmethod
+    def exists(self):
+        """Checks whether the file exists"""
+
+    @abstractmethod
+    def to_input_file(self):
+        """Returns an InputFile for the location of this output file"""
+
+    @abstractmethod
+    def __enter__(self, overwrite: bool = False):
+        """Enter context for OutputFile
+
+        This method should create a new file and set an output stream to `self.output_stream`
+        and return self. If the file already exists and `overwrite` is False, a FileExistsError
+        should be raised."""
+
+        log.info(f"Opening file {self._location}")
+
+    @abstractmethod
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        """Exit context for OutputFile
+
+        This method should perform any necessary teardown. To ensure that closing the
+        file is logged, call `super().__exit__()` at the beginning of the implementation
+        of this method."""
+
+        log.info(f"Closing file {self._location}")
```

Comment:
I don't think that the abstract class should handle logging. Is this an existing convention?